pkg/storer/compact.go

```
@@ -0,0 +1,192 @@
+					}
+
+					from.Location = to
+					if err := store.Put(from); err != nil {
+						return fmt.Errorf("store put: %w", err)
+					}
+
+					start++
+					end--
+				} else {
+					end-- // keep moving to the left until a used slot is found
+				}
+			} else {
+				start++ // keep moving to the right until a free slot is found
+			}
+		}
+
+		logger.Info("shard truncated", "shard", fmt.Sprintf("%d/%d", shard, sharkyNoOfShards-1), "slot", end)
+
+		if err := sharkyRecover.TruncateAt(context.Background(), uint8(shard), end+1); err != nil {
+			return fmt.Errorf("sharky truncate: %w", err)
+		}
+	}
+
+	if err := sharkyRecover.Save(); err != nil {
+		return fmt.Errorf("sharky save: %w", err)
+	}
+
+	logger.Info("compaction finished", "duration", time.Since(n))
+
+	if validate {
+		logger.Info("performing chunk validation after compaction")
+		validationWork(logger, store, sharkyRecover)
+	}
+
+	return nil
+}
+
+func validationWork(logger log.Logger, store storage.Store, sharky *sharky.Recovery) {
+
+	n := time.Now()
+	defer func() {
+		logger.Info("validation finished", "duration", time.Since(n))
+	}()
+
+	iteratateItemsC := make(chan *chunkstore.RetrievalIndexItem)
+
+	validChunk := func(item *chunkstore.RetrievalIndexItem, buf []byte) error {
+		err := sharky.Read(context.Background(), item.Location, buf)
+		if err != nil {
+			return err
+		}
+
+		ch := swarm.NewChunk(item.Address, buf)
+		if !cac.Valid(ch) && !soc.Valid(ch) {
+			return errors.New("invalid chunk")
+		}
+
+		return nil
+	}
+
+	eg := errgroup.Group{}
+
+	for i := 0; i < 8; i++ {
+		eg.Go(func() error {
+			buf := make([]byte, swarm.SocMaxChunkSize)
+			for item := range iteratateItemsC {
+				if err := validChunk(item, buf[:item.Location.Length]); err != nil {
+					logger.Info("invalid chunk", "address", item.Address, "error", err)
+				}
+			}
+			return nil
+		})
+	}
+
+	count := 0
+	_ = chunkstore.Iterate(store, func(item *chunkstore.RetrievalIndexItem) error {
```

you ignoring the errors then you should not use errgroup but waitgroup